x-pack/plugins/ml/public/application/jobs/new_job/job_from_lens/create_job.ts

```diff
@@ -0,0 +1,249 @@
+    stashJobForCloning(
+      {
+        jobConfig,
+        datafeedConfig,
+        createdBy,
+        start,
+        end,
+      } as JobCreatorType,
+      true,
+      includeTimeRange,
+      !includeTimeRange
+    );
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.error(error);
+  }
+}
+
+export async function canCreateADJob(
+  vis: LensSavedObjectAttributes,
+  query: Query | undefined,
+  filters: Filter[] | undefined,
+  dataViewClient: DataViewsContract,
+  kibanaConfig: IUiSettingsClient
+) {
+  try {
+    const jobItems = await createADJobFromLensSavedObject(
+      vis,
+      query ?? getDefaultQuery(),
+      filters ?? [],
+      dataViewClient,
+      kibanaConfig
+    );
+
+    return !!jobItems;
+  } catch (error) {
+    return false;
+  }
+}
+
+async function createADJobFromLensSavedObject(
+  vis: LensSavedObjectAttributes,
+  query: Query,
+  filters: Filter[],
+  dataViewClient: DataViewsContract,
+  kibanaConfig: IUiSettingsClient
+) {
+  const { fields, timeField, splitField, dataView } = await extractFields(vis, dataViewClient);
+
+  const jobConfig = createEmptyJob();
+  const datafeedConfig = createEmptyDatafeed(dataView.title);
+
+  const mainQuery = query?.query !== '' ? query : vis.state.query;
+
+  const { combinedQuery } = createQueries(
+    { query: mainQuery, filter: filters },
+    dataView,
+    kibanaConfig
+  );
+  datafeedConfig.query = combinedQuery;
+
+  jobConfig.analysis_config.detectors = createDetectors(fields, splitField);
+
+  jobConfig.data_description.time_field = timeField.sourceField;
+  jobConfig.analysis_config.bucket_span = '15m';
+  if (splitField) {
+    jobConfig.analysis_config.influencers = [splitField.sourceField];
+  }
+
+  const createdBy =
+    splitField || jobConfig.analysis_config.detectors.length > 1
+      ? CREATED_BY_LABEL.MULTI_METRIC
+      : CREATED_BY_LABEL.SINGLE_METRIC;
+
+  return {
+    jobConfig,
+    datafeedConfig,
+    createdBy,
+  };
+}
+
+async function extractFields(vis: LensSavedObjectAttributes, dataViewClient: DataViewsContract) {
+  const visualization = vis.state.visualization as { layers: XYDataLayerConfig[] };
+  const indexpattern = vis.state.datasourceStates.indexpattern as IndexPatternPersistedState;
+
+  const compatibleLayers = visualization.layers.filter(
+    ({ layerType, seriesType }) =>
+      layerType === COMPATIBLE_LAYER_TYPE && COMPATIBLE_SERIES_TYPES.includes(seriesType)
+  );
+
+  const [firstCompatibleLayer] = compatibleLayers;
+
+  const compatibleIndexPatternLayer = Object.entries(indexpattern.layers).find(
+    ([id]) => firstCompatibleLayer.layerId === id
+  );
+  if (compatibleIndexPatternLayer === undefined) {
+    throw Error(
+      'Visualization does not contain any layers which can be used for creating an anomaly detection job.'
+    );
+  }
+
+  const [layerId, layer] = compatibleIndexPatternLayer;
+
+  const { columns } = layer as { columns: Record<string, FieldBasedIndexPatternColumn> };
```

Comment:
> Maybe on this level it should be filtered a bit.

I experimented with filtering the fields first and decided it was unnecessary as the error thrown from `sourceField` not existing means the vis is incompatible and so the menu item is not shown.
Relying on an error like that is a bit hacky, so I'll add a filter and throw a nicer error.

> there is formula, moving average and so on.

we can only support a subset of operation types, formula and moving average cannot be supported by our detectors.
Unless the operation is `average`, `count`, `max`, `median`, `min`, `sum` or `unique_count` the menu item will not be supported.


> Also, columns might have a time shift and/or a "filter"

I completely missed this in the UI!?!
I don't think we can support either of these features for ML job detectors.
Filters might be possible by using aggregations in the datafeed, but it would need investigation.
For this first version of the feature, I think columns with a `filter` or `timeShift` should not be allowed. And if they are being used, the whole vis should be incompatible.

Because there are subtle ways a vis can be deemed incompatible, I think it would be good to let the user know the reasons why.
At the moment we just hide the link for incompatible visualisations, but perhaps we should instead always show the link and have it open a flyout to show analysis of whether it can be used for a job.
This could be broken down by layer and allow the user to pick which layer to use for the ML job.

Reply:
It's in the "add advanced options": 
<img width="435" alt="Screenshot 2022-05-11 at 16 47 58" src="https://user-images.githubusercontent.com/1508364/167878796-30794ef8-1ce3-4b59-8ff0-ce91d57f0150.png">

Should be easy to filter for these, they are just properties on the column

Reply:
I've added checks for `filter` and `timeShift` [here](https://github.com/elastic/kibana/pull/129762/commits/a382a3ff587e4c72ed99fded97851088a0090ea9)
If any columns contain them, the `Create Anomaly Detection Job` menu item is not displayed.

As I mentioned above, it would be much nicer to inform the user why that visualisation cannot be used for a AD job as it won't be obvious.
Hopefully I'll have time to add this information flyout for 8.3.